Update JS Learn Events page

wbamberg opened the pull request:
Part of https://github.com/mdn/content/issues/10337.

The main thing here is to be consistent about describing `addEventListener()` as the recommended way to listen for events.

Before this the page did recommend `addEventListener()`, but also described all three mechanisms together and all the examples used event handler properties. In this update we describe `addEventListener()` first, then the other two as alternatives, and the examples should consistently use `addEventListener()`.

I also made some structural changes, quite a few copy edits, and made all the code samples use template literals and `const` where appropriate.

This PR will need a companion PR to https://github.com/mdn/learning-area, which I'll work on next.

wbamberg committed:
Update JS Learn Events page

wbamberg committed:
Some more event updates

Elchi3 commented on files/en-us/learn/javascript/building_blocks/events/index.md:
```diff
@@ -410,12 +407,34 @@ div {
+Most event objects have a standard set of properties and methods available on the event object; see the {{domxref("Event")}} object reference for a full list.
 
-### Preventing default behavior
+Some event objects add extra properties that are relevant to that particular type of event. For example, the {{domxref("Document/keydown_event", "keydown")}} event fires when the user presses a key, and its event object has a `key` property that tells you which key was pressed:
```
mention here that `key` comes from https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent which is a specialized `Event` object?

Elchi3 commented on files/en-us/learn/javascript/building_blocks/events/index.md:
```diff
@@ -126,225 +128,223 @@ See the [`runtime.onMessage`](/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runt
 
-In addition, there are other powerful features and options available with this event mechanism.
-These are a little out of scope for this article, but if you want to read them, visit the [`addEventListener()`](/en-US/docs/Web/API/EventTarget/addEventListener) and [`removeEventListener()`](/en-US/docs/Web/API/EventTarget/removeEventListener) reference pages.
+> **Note:** Separating your programming logic from your content also makes your site more friendly to search engines.
```
Not sure if this is a thing learners are concerned with at stage, but another point would be that CSP often also disallows the inline JS code.

dontcallmedom replied:
without diving into CSP, maybe a mention that there are servers where it be blocked from executing, as an extra motivation?

Elchi3 commented on files/en-us/learn/javascript/building_blocks/events/index.md:
```diff
@@ -126,225 +128,223 @@ See the [`runtime.onMessage`](/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runt
+
+There are many different events that can be fired by a button element. Let's experiment.
 
 First, make a local copy of [random-color-eventhandlerproperty.html](https://github.com/mdn/learning-area/blob/master/javascript/building-blocks/events/random-color-eventhandlerproperty.html), and open it in your browser.
```
```suggestion
First, make a local copy of [random-color-addeventlistener.html](https://mdn.github.io/learning-area/javascript/building-blocks/events/random-color-addeventlistener.html), and open it in your browser.
```

Elchi3 reviewed (approved):
Nice work! Just some nits

dontcallmedom commented on files/en-us/learn/javascript/building_blocks/events/index.md:
```diff
@@ -61,9 +63,9 @@ For example:
+You can gather from this (and from glancing at the MDN [event reference](/en-US/docs/Web/Events)) that there are **a lot** of events that can be responded to.
 
-Each available event has an **event handler**, which is a block of code (usually a JavaScript function that you as a programmer create) that runs when the event fires.
+To respond to an event, you attach an **event handler** to it. This is a block of code (usually a JavaScript function that you as a programmer create) that runs when the event fires.
```
```suggestion
To react to an event, you attach an **event handler** to it. This is a block of code (usually a JavaScript function that you as a programmer create) that runs when the event fires.
```
(nit)

wbamberg committed:
typo fix - double period

Co-authored-by: Florian Scholz <fs@florianscholz.com>

wbamberg committed:
Fix link to GitHub

Co-authored-by: Florian Scholz <fs@florianscholz.com>

wbamberg committed:
respond->react

Co-authored-by: Dominique Hazael-Massieux <dom@w3.org>

wbamberg committed:
Review comments updates

wbamberg commented:
Thanks for the reviews! I just pushed https://github.com/mdn/content/pull/10914/commits/e0ea26b51f95a5d8d4efde07286d8e416dd5fdd4 which:

* tries to be more consistent about using "react" rather than "respond" (I guess the thinking here is that "respond" might make people think of sending an answer back, and I agree with that).
* adds a bit about servers potentially disallowing inline JS
* links to `KeyboardEvent`

The more I look at the event bubbling explanation though the less I like it. "Event bubbling and capture are two mechanisms that describe what happens when two handlers of the same event type are activated on one element" seems misleading to me. So I'm going to try reworking that a little, and also move the "event target" example down to event delegation.

wbamberg committed:
More info on event bubbling; move and modify event target example

wbamberg commented:
I just pushed https://github.com/mdn/content/pull/10914/commits/7cdf6db0c5aae115e8b2b7d61e51029850641e29, which adds a more gentle, layered intro to event bubbling. This is not an easy concept and I felt the video-hiding example was a bit in at the deep end. I hope you agree this is better!

I also moved the event.target example to the event delegation section, and updated it to use event delegation.

@Elchi3 , ready for re-review.

hamishwillee commented on files/en-us/learn/javascript/building_blocks/events/index.md:
```diff
@@ -126,225 +128,229 @@ See the [`runtime.onMessage`](/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runt
+Finally, many common server configurations will disallow inline JavaScript, as a security measure.
 
-### What mechanism should I use?
+## Which mechanism should I use?
```
@wbamberg This is all an excellent improvement - in particular making `addEventListener` first and foremost.

FWIW I would remove this section, integrating the key point under the heading  "Other event listener mechanisms" and in each of the parts. 

So in "Other event listener mech" - neither of these methods are recommended as the addEventListener approach is easier and safer to use. Then in the onXX section explain that the limitation that it can only be used once.

I know this is a preference, but I think it is better to introduce the alternatives as options you should not use rather than peers and then explaining later that they aren't really.

wbamberg replied:
Yeah, I agree with this - that section was really a remnant of the previous organization. I just pushed https://github.com/mdn/content/pull/10914/commits/5d4d29c963188ccadb56d10798c0ac4d311e77e1 which removes it.

hamishwillee replied:
Good fix. I won't add an approval while others are still doing detail review, but the whole thing is much better (and it already wasn't "bad").

wbamberg committed:
Remove 'which should you use' section

wbamberg committed:
wording

dontcallmedom commented on files/en-us/learn/javascript/building_blocks/events/index.md:
```diff
@@ -474,14 +444,96 @@ The output is as follows:
-Event bubbling and capture are two mechanisms that describe what happens when two handlers of the same event type are activated on one element.
+## Event bubbling and capture
+
+Event bubbling and capture are terms that describe how the browser handles events when elements are nested inside each other.
```
```suggestion
Event bubbling and capture are terms that describe phases in how the browser handles events targeted at nested elements.
```
(except for the root, elements are pretty much always nested, so "when elements are nested" isn't very informative)

wbamberg committed:
Update files/en-us/learn/javascript/building_blocks/events/index.md

Co-authored-by: Dominique Hazael-Massieux <dom@w3.org>

Elchi3 reviewed (approved):
Great work! 👍
